quinn-proto/src/connection/streams/recv.rs

```diff
@@ -226,37 +195,105 @@ impl Recv {
     }
 }
 
-pub(crate) type ReadResult<T> = Result<Option<DidRead<T>>, ReadError>;
-
-/// Result of a `Streams::read` call in case the stream had not ended yet
-#[derive(Debug, Eq, PartialEq, Copy, Clone)]
-#[must_use = "A frame might need to be enqueued"]
-pub(crate) struct DidRead<T> {
-    pub result: T,
-    pub max_stream_data: ShouldTransmit,
-    pub max_data: ShouldTransmit,
+/// An iterator-like interface to read data from a receive stream
+pub struct Chunks<'a> {
+    id: StreamId,
+    streams: &'a mut Streams,
+    pending: &'a mut Retransmits,
+    recv: Option<Recv>,
+    ordered: bool,
+    read: u64,
+    blocked: bool,
 }
 
-pub(super) type StreamReadResult<T> = Result<Option<T>, ReadError>;
+impl<'a> Chunks<'a> {
+    pub(crate) fn new(
+        id: StreamId,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+        ordered: bool,
+    ) -> Result<Self, ReadError> {
+        let mut entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadError::UnknownStream),
+        };
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
-}
+        let rs = entry.get_mut();
+        let recv = if !rs.assembler.can_read(ordered)? {
+            match rs.read_blocked() {
+                Ok(()) => {
+                    entry.remove_entry();
+                    streams.stream_freed(id, StreamHalf::Recv);
+                    None
+                }
+                Err(e @ ReadError::Reset { .. }) => {
+                    entry.remove_entry();
+                    streams.stream_freed(id, StreamHalf::Recv);
+                    return Err(e);
+                }
+                Err(e) => return Err(e),
+            }
+        } else {
+            Some(entry.remove())
+        };
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+        Ok(Self {
+            id,
+            recv,
+            streams,
+            pending,
+            ordered,
+            read: 0,
+            blocked: false,
+        })
     }
-}
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+    /// Read the next chunk from the stream
+    ///
+    /// Return the next chunk of data from the stream, if any are available. Will return `None` if
+    /// all available data has been read. `max_length` may be used to restrict the maximum size
+    /// of the returned `Bytes` (but `usize::MAX` is supported, too).
+    ///
+    /// Will keep returning `None` once no further data is available for reading.
+    pub fn next(&mut self, max_length: usize) -> Option<Chunk> {
```

Comment:
If we moved `max_length` into `self`, we could implement `Iterator` in earnest, and also make it easier for the caller to limit the *total* length rather than the length of an individual chunk (which admittedly doesn't seem important for our current use, but is IMO a nicer API in general).

Reply:
I did consider this, but passing `max_length` on every iteration seemed more flexible (the caller could decide to change the `max_length` in further iterations) and requires no extra complexity on the part of the assembler.

Reply:
Fair enough; no strong feeling.